### ISSUE TITLE
Combine all three theme types into one

### ISSUE THREAD

**Issue by karlhorky** [saasquatch/time-credit-theme#19]
It would be nice to combine all three theme types ([time-credit](https://github.com/saasquatch/time-credit-theme), [dollar-credit](https://github.com/saasquatch/dollar-credit-theme), [percent-credit](https://github.com/saasquatch/percent-credit-theme)) into one repo.

We could do this with a similar trick as #2 - by:
- moving the required content and components from the other repos into this one
- setting another container class with the `rewardBalances.type` value
- showing and hiding the different content and components based on that class

@loganvolkers @torbensky any reason why we don't want to combine all the three themes into one single repo?

**Comment by cloud9yupiq** [comment]
I don't know any reason why we wouldn't want these all in the same repo.

The only benefit of our current setup would seem to be independent
versioning... And easier forking for people writing custom themes.

**Comment by karlhorky** [comment]
Right, yeah that's what I was thinking too.

I wonder if we could make three extra branches in the main repository and create an automated process like a post-commit hook to update these branches when changes were committed to master. Then we could get the best of both worlds.

**Comment by cloud9yupiq** [comment]
Ability to easily fork and custom version probably isn't worth it.

Maybe easier to just have a separate starter theme to fork for custom
themes...

Vanilla theme are prone to have a lot of levers and knobs in them that
makes them fundamentally more complicated and harder to fork.

Single repo seems better for maintainability.

Right, yeah that's what I was thinking too.

I wonder if we could make three extra branches in the main repository and
create an automated process like a post-commit hook to update these
branches when changes were committed to master. Then we could get the best
of both worlds.

—
Reply to this email directly or view it on GitHub
<https://github.com/saasquatch/time-credit-theme/issues/19#issuecomment-75457751>
.

**Comment by karlhorky** [comment]
Sounds good, then I'll take a look at what it would take to merge them.

**Comment by torbensky** [comment]
I don't see a strong motivation to separate or merge them, so I think we might as well keep it the way it is. Seems to me that the current method is working, so I don't understand the reasoning for devoting time into this over other priorities. I would say we shouldn't do this unless clear value can be demonstrated.

**Comment by karlhorky** [comment]
@torbensky the value is in the reduced maintenance time if we do the merge. Right now every feature or bug fix that affects all three themes (the normal case) takes three times as long to implement, and has an increased chance of error. With a single repo these effects are reduced.

**Comment by torbensky** [comment]
@karlhorky That makes sense. However, the flip side of that is we have less flexibility to change each style of template and have to spend more effort on building generic components that can be adapted to either case. It also increases the complexity of understanding the templates for 3rd parties. Not sure what the overall trade-off is.

**Comment by karlhorky** [comment]
This may be accomplished through the technique demonstrated in saasquatch/base-theme (using the [dynamic partials functionality](https://github.com/jknack/handlebars.java/issues/375) available in Handlebars.java 2.2).

**Comment by karlhorky** [comment]
Now that we have `2.2.0`, all of what we originally needed to merge these repos has been implemented.

However, while waiting for `2.2.0`, the theme variables were implemented, which define a structure for setting messaging and other variables for a specific theme and program type. The theme variables also define default values for the messages, such as "Get x months of free!". This is tightly coupled to the program type, which would cause the same default message to be set for all program types of a theme. I took a look and I think I've come up with a solution to decouple the theme variables from the program type.

With [`$ref` and `allOf`](http://spacetelescope.github.io/understanding-json-schema/structuring.html#extending) we can create mixins for our variables. In this way we can create multiple definitions for the default values per program type.

I've created a demo here: https://gist.github.com/karlhorky/497389befc9919805d42

Again, the benefits of doing this:
- reduces 3 repos to one, increasing ease of use
- reduces code duplication across 3 repos
- updates to themes can be completed in 1/3 the time
​
I do see potential for forgetting to add a defaults definition for the program type. The danger of forgetting to add a defaults key is however also present without this change, so I'm not sure this adds much risk. I also think that the benefits above outweigh this added complexity.

@jericmason @torbensky @loganvolkers thoughts?

**Comment by karlhorky** [comment]
I do see potential for forgetting to add a defaults definition for the program type. However, JSON Schema is generally not so tolerant of forgetting to add something, so this is not really a new problem.

**Comment by torbensky** [comment]
This sounds pretty good overall. I need to do some testing with the JSON Schema stuff to see if that will work. AFAIK we are using Draft 4, but we may not have implemented everything on our backend to support handling things like $ref properly.

**Comment by karlhorky** [comment]
Yeah @loganvolkers mentioned this would need some backend changes - @jericmason made a ticket for this yupiq/saasquatch#460 .

**Comment by loganvolkers** [comment]
I believe that Line 28 of Karl's demo is a technical blocker: https://gist.github.com/karlhorky/497389befc9919805d42#file-variables-json-L28

There's no `if` or `switch` statements that I know of in JSON Schema.

Not to say that it's impossible to somehow have a vanilla theme that allows for multi and mixed reward types, but just the `allOf` approach proposed on Line 28 does not seem possible.
